\begin{verbatim}
./kallisto index -i flax_ref.index SRR11830041.1_1.fasta

[build] loading fasta file SRR11830041.1_1.fasta
[build] k-mer length: 31
[build] warning: clipped off poly-A tail (longer than 10)
from 997 target sequences
[build] warning: replaced 81307 non-ACGUT characters in the input sequence
with pseudorandom nucleotides
[build] counting k-mers ... done.
[build] building target de Bruijn graph ...  done
[build] creating equivalence classes ...  done
[build] target de Bruijn graph has 12250208 contigs and contains 106849682 k-mers

\end{verbatim}

\begin{verbatim}
./kallisto quant -i flax_ref.index -o paired_reads/out/SRR8177761 paired_reads/SRR8177761_1_paired.fastq.gz  paired_reads/SRR8177761_2_paired.fastq.gz


[quant] fragment length distribution will be estimated from the data
[index] k-mer length: 31
[index] number of targets: 8,078,958
[index] number of k-mers: 106,849,682
[index] number of equivalence classes: 16,334,370
[quant] running in paired-end mode
[quant] will process pair 1: paired_reads/SRR8177761_1_paired.fastq.gz
paired_reads/SRR8177761_2_paired.fastq.gz
[quant] finding pseudoalignments for the reads ... done
[quant] processed 8,427,058 reads, 338,347 reads pseudoaligned
[quant] estimated average fragment length: 125.696
[   em] quantifying the abundances ... done
[   em] the Expectation-Maximization algorithm ran for 652 rounds
\end{verbatim}

\subsection{STAR}

\begin{verbatim}
/mnt/e/opt/STAR-2.7.10a/bin/Linux_x86_64_static # ./STAR --runThreadN 40 --runMode genomeGenerate --genomeFastaFiles ../../../../../y/vzuev/datasets/bioMG2022/lus.final.fasta --sjdbGTFfile ../../../../../y/vzuev/ # regularly tap ENTER

./STAR --runThreadN 40 --genomeDir GenomeDir/ --readFilesIn \
> ../../../../../y/vzuev/datasets/bioMG2022/SRX4997855/SRR8177762_1_paired.fastq \
> ../../../../../y/vzuev/datasets/bioMG2022/SRX4997855/SRR8177762_2_paired.fastq

./STAR --runThreadN 40 --genomeDir GenomeDir/ --readFilesIn \
> ../../../../../y/vzuev/datasets/bioMG2022/SRX4997855/SRR8177762_1_paired.fastq \
> ../../../../../y/vzuev/datasets/bioMG2022/SRX4997855/SRR8177762_2_paired.fastq \
> --sjdbInsertSave All  --quantMode GeneCounts --outReadsUnmapped Fastq --outSAMstrandField intronMotif

tail -n +5 ReadsPerGene.out.tab > gene_counts.count

# mass production: # https://github.com/najasplus/STAR-deseq2

for i in 44 55 56 57 58; do
    ./STAR --runThreadN 40 --genomeDir GenomeDir/ --readFilesIn \
        "SRR81777${i}_1_paired.fastq" \
        "SRR81777${i}_2_paired.fastq" \
        --sjdbInsertSave All  --quantMode GeneCounts --outReadsUnmapped Fastq --outSAMstrandField intronMotif \
        --outFileNamePrefix "./SRR81777${i}/"
done

for i in 55 56 57 58 59 60 61 62; do
    cd "./SRR81777${i}/"
    tail -n +5 ReadsPerGene.out.tab > gene_counts.count
    cd ..
done

\end{verbatim}

\begin{verbatim}
# https://data.jgi.doe.gov/refine-download/phytozome?genome_id=200&expanded=Phytozome-200

root@powerhouse:/mnt/y/vzuev/datasets/bioMG2022/phytozome# /mnt/e/opt/STAR-2.7.10a/bin/Linux_x86_64_static/STAR --runThreadN 40 --runMode genomeGenerate --genomeFastaFiles Lusitatissimum_200_BGIv1.0.fa --sjdbGTFfile Lusitatissimum_200_v1.0.gene_exons.gtf --limitGenomeGenerateRAM=100000000000

/mnt/e/opt/STAR-2.7.10a/bin/Linux_x86_64_static/STAR --runThreadN 40 --runMode genomeGenerate --genomeFastaFiles Lusitatissimum_200_BGIv1.0.fa --sjdbGTFfile Lusitatissimum_200_v1.0.gene_exons.gtf --limitGenomeGenerateRAM=100000000000
STAR version: 2.7.10a   compiled: 2022-01-14T18:50:00-05:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
Jul 04 23:41:07 ..... started STAR run
Jul 04 23:41:07 ... starting to generate Genome files
Jul 04 23:41:53 ..... processing annotations GTF
!!!!! WARNING: --genomeSAindexNbases 14 is too large for the genome size=318250901, which may cause seg-fault at the mapping step. Re-run genome generation with recommended --genomeSAindexNbases 13
Jul 04 23:42:40 ... starting to sort Suffix Array. This may take a long time...
Jul 04 23:43:26 ... sorting Suffix Array chunks and saving them to disk...
Jul 04 23:46:41 ... loading chunks from disk, packing SA...
Jul 04 23:47:09 ... finished generating suffix array
Jul 04 23:47:09 ... generating Suffix Array index
Jul 04 23:48:21 ... completed Suffix Array index
Jul 04 23:48:21 ... writing Genome to disk ...
Jul 05 00:02:58 ... writing Suffix Array to disk ...
Jul 05 00:04:40 ... writing SAindex to disk
Jul 05 00:05:44 ..... finished successfully


\end{verbatim}
