---
title: "GWAS var1"
author: Valerii Zuev
date: 12/23/2022
output: html_notebook
---
## 2. Загрузить генетические и клинические данные

Necessary libraries:
```{r message = F, warning = F}
library(gdsfmt) # internal R format for genetic data
library(SNPRelate)
library(vcfR)
library(HardyWeinberg)
setwd(here::here("r/stats2022fall/VAR_1")) # TODO not when HTML generated
```

Converting VCF to GDS:
```{r}
vcf.fn <- "genotype1.vcf.gz"
pht.fn <- "phenotype_1.txt"
gds.fn <- "DATA.gds"

# snpgdsVCF2GDS(vcf.fn, gds.fn) # call this once to create a GDS file
```

Reading elements:
```{r}
dat.f <- snpgdsOpen(gds.fn)
sample.id <- read.gdsn(index.gdsn(dat.f, "sample.id"))
snp.id <- read.gdsn(index.gdsn(dat.f, "snp.id"))
snp.rs.id <- read.gdsn(index.gdsn(dat.f, "snp.rs.id"))
chr <- read.gdsn(index.gdsn(dat.f, "snp.chromosome"))
pos <- read.gdsn(index.gdsn(dat.f, "snp.position"))
allele <- read.gdsn(index.gdsn(dat.f, "snp.allele"))
genotype <- read.gdsn(index.gdsn(dat.f, "genotype"))
snpgdsClose(dat.f)
```

Adding annotation:
```{r}
pht <- read.table(pht.fn, head = T)
dat.f.writeable <- snpgdsOpen(gds.fn, readonly = F)
add.gdsn(dat.f.writeable, "sample.annot", val = pht, compress = "ZIP", closezip = TRUE) # TODO not after the first read
snpgdsClose(dat.f.writeable)
```

Opening GDS file again and reading annotation (clinical data):
```{r}
dat.f <- snpgdsOpen(gds.fn)
sample.annot <- read.gdsn(index.gdsn(dat.f, "sample.annot"))
snpgdsClose(dat.f)
```

## 3. Охарактеризовать имеющиеся клинические данные. Определить частоты экспериментальной и контрольной групп в мужской и женской популяциях. При наличии пропущенных значений наблюдаемой переменной удалить из исследования соответствующих индивидов.

```{r}
cat("Gender levels: ")
levels(as.factor(sample.annot$sex)) %>% cat("\n")

cat("Phenotype levels (2 - HIV-infected, 1 - control): ")
levels(as.factor(sample.annot$phenotype)) %>% cat("\n")

cat("\nTotal number of observations: ", nrow(sample.annot), "\n\n")

cat("Observations of case/control by gender\n")
table(sample.annot$sex, sample.annot$phenotype)

cat("\nMissing phenotype values:\n")
is.na(sample.annot$sex) %>% sum
is.na(sample.annot$phenotype) %>% sum
```
## 4. Охарактеризовать имеющиеся генетические данные. Вычислить для каждого генотипа: частоту минороной аллели (MAF); коэффициент гетерозиготности; процент генотипированных образцов (call rate).

```{r}
# what's inside `genotype`?
cat("number of samples:", nrow(genotype))
cat("number of SNPs:", ncol(genotype))
genotype[1:3, 1:10]
sum(genotype == 3) # >40 000
# 0 <=> no SNP,  1 <=> SNP in one allele, 2 <=> SNP in both alleles, 3 <=> missing value

genotype_with_na <- genotype
genotype_with_na[genotype == 3] <- NA
genotype_observations <- colSums(!is.na(genotype_with_na))
maf <- 1 - (colSums(genotype_with_na, na.rm = T) / 2) / genotype_observations
# maf <- lapply(maf, function (x) if(x <= .5) x else 1 - x) %>% unlist
head(maf)
```

Heterozygosity

```{r}
heterozygocity <- colSums(genotype == 1) / colSums(genotype, na.rm = T) # this is probably correct
expected.heterozygocity <- 2 * maf * (1 - maf)

head(heterozygocity)
head(expected.heterozygocity)
hist(heterozygocity)
hist(expected.heterozygocity)
hist(expected.heterozygocity - heterozygocity)

```

Call rate

```{r}
cr <- genotype_observations / nrow(genotype)
min(cr)
```

## 5. Провести исследование качества данных для поиска генетических ассоциаций. Отфильтровать данные, не удовлетворяющие стандартам качества.

Использовать фильтры:
- число пропущенных значений наблюдаемой переменной > 2%
- число негенотипированных индивидов > 5%
- MAF < 0.05
- p-значение теста на проверку равновесия Харди–Вайнберга (pv < 10^{−5} в контрольной группе
- pv < 10^{−10} в группе заражённых

```{r}
min.callrate <- .95 # minimum fraction of observations per SNP
min.genotypedSNP <- .98 # minimum fraction of genotyped SNPs per observation
min.maf <- .05 # minimum MAF per SNP

crFilter <- cr < min.callrate
mafFilter <- maf < min.maf

genotype.rowCounts <- rowSums(!is.na(genotype_with_na))
individFilter <- genotype.rowCounts < min.genotypedSNP

sum(crFilter) # 815
sum(mafFilter & !crFilter) # 34479
sum(mafFilter & crFilter) # 157; 157/34479 [almost equals] 815 / length(cr)
sum(individFilter) # 0
```
HWE test
```{r}

create_freq_mtx <- function(phenotype_value = ? float) {
  gen.subset <- genotype[sample.annot$phenotype == phenotype_value,]
  ant.subset.df <- data.frame(AA = colSums(gen.subset == 0), AB = colSums(gen.subset == 1), BB = colSums(gen.subset == 2))
  return(as.matrix(ant.subset.df))
} -> matrix

hwe.subset.f <- function(i = ? integer, subset.mtx = ? matrix) {
  HWExact(subset.mtx[i,], verbose = F)$pval
}

control.freq <- create_freq_mtx(1)
case.freq <- create_freq_mtx(2)

case.hwe <- NULL
control.hwe <- NULL

for (i in seq_along(chr)) {
  case.hwe <- c(case.hwe, hwe.subset.f(i, case.freq))
  control.hwe <- c(control.hwe, hwe.subset.f(i, control.freq))
}
# save.image(file="5a.RData") # saving checkpoint
```

Putting together all filters:
```{r}
# don't forget to load libraries and change path!
load(file = "5a.RData") # restoring from checkpoint

hweCaseFilter <- case.hwe < 10^(-10)
hweControlFilter <- control.hwe < 10^(-5)
sum(hweCaseFilter) # 0
sum(hweControlFilter) # 1
sum(hweControlFilter & !(mafFilter | crFilter)) # 1
allFilter <- !(crFilter | mafFilter | hweControlFilter)

snp.id <- snp.id[allFilter]
chr <- chr[allFilter]
pos <- pos[allFilter]
allele <- allele[allFilter]
genotype <- genotype[, allFilter]

```

## 6. Провести анализ главных компонент. При наличии в популяции нетипичных образцов отфильтровать их. Интерпретировать результат

Separate GDS file for PCA:
```{r}
dat.f <- createfn.gds("DATA_PCA.gds")
add.gdsn(dat.f, "sample.id", val = sample.id, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.id", val = snp.id, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.chromosome", val = chr, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.position", val = pos, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.allele", val = allele, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "genotype", val = genotype, compress = "ZIP", closezip = TRUE, storage = "bit2", replace = TRUE)
add.gdsn(dat.f, "sample.annot", val = sample.annot, compress = "ZIP", closezip = TRUE)
put.attr.gdsn(index.gdsn(dat.f, "genotype"), "sample.order")
closefn.gds(dat.f)
```

```{r}
set.seed(0) # for reproducibility
dat.f <- snpgdsOpen("DATA_PCA.gds")
ftr <- snpgdsLDpruning(dat.f) # filtering by LD (linkage disequilibrium)
ftr.id <- unlist(ftr)
q <- snpgdsPCA(dat.f, snp.id = ftr.id)
id <- read.gdsn(index.gdsn(dat.f, "sample.id"))
ant <- read.gdsn(index.gdsn(dat.f, "sample.annot"))
snpgdsClose(dat.f)
```
Checkpoint:
```{r}
# save.image(file="6a_pca.RData")
load("6a_pca.RData")
```
First two components (color <=> gender: blue <=> female):
```{r}
vec <- q$eigenvect
x <- vec[, 1]
y <- vec[, 2]
z <- vec[, 3]

pairs(data.frame(x = x, y = y, z = z))


plot(x, y, "n")
points(x[ant$sex == "F"], y[ant$sex == "F"], col = "blue")
points(x[ant$sex == "M"], y[ant$sex == "M"], col = "red")

```
First two components + case/control:
```{r}
smp <- ant$phenotype
clr <- c("blue", "red")
smp.l <- c("Controls", "Cases")
pc.prt <- q$varprop * 100
lbs <- paste0("PC", 1:2, "(", format(pc.prt[1:2], digits = 2), "%)")

plot(x, y, "n", xlab = lbs[1], ylab = lbs[2])
points(x[ant$phenotype == 1], y[ant$phenotype == 1], col = "blue")
points(x[ant$phenotype == 2], y[ant$phenotype == 2], col = "red")
legend("topright", legend = smp.l, col = clr, pch = 1, cex = 1)

```
Filtering:
```{r}
pcaFilter <- (x > -.15) & (y < .1) & (z < .1)
sum(!pcaFilter) # 9

pht <- pht[pcaFilter,]

# dat.f <- createfn.gds("DATA_PCA2.gds")
# add.gdsn(dat.f, "sample.id", val = id[pcaFilter], compress = "ZIP", closezip = TRUE)
# add.gdsn(dat.f, "snp.id", val = snp.id, compress = "ZIP", closezip = TRUE)
# add.gdsn(dat.f, "snp.chromosome", val = chr, compress = "ZIP", closezip = TRUE)
# add.gdsn(dat.f, "snp.position", val = pos, compress = "ZIP", closezip = TRUE)
# add.gdsn(dat.f, "snp.allele", val = allele, compress = "ZIP", closezip = TRUE)
# add.gdsn(dat.f, "genotype", val = genotype[pcaFilter,], compress = "ZIP", closezip = TRUE, storage = "bit2", replace = TRUE)
# add.gdsn(dat.f, "sample.annot", val = ant[pcaFilter,], compress = "ZIP", closezip = TRUE)
# put.attr.gdsn(index.gdsn(dat.f, "genotype"), "sample.order")
# closefn.gds(dat.f)
```

PCA again:
```{r}
# TODO make a function? (code repetition)
set.seed(0)
dat.f <- snpgdsOpen("DATA_PCA2.gds")
ftr <- snpgdsLDpruning(dat.f)
ftr.id <- unlist(ftr)
q <- snpgdsPCA(dat.f, snp.id = ftr.id)
id <- read.gdsn(index.gdsn(dat.f, "sample.id"))
ant <- read.gdsn(index.gdsn(dat.f, "sample.annot"))
snpgdsClose(dat.f)
# pair
vec <- q$eigenvect
x2 <- vec[, 1]
y2 <- vec[, 2]
z2 <- vec[, 3]
```
```{r}
plot(x2, y2, "new")
points(x2[ant$sex == "F"], y2[ant$sex == "F"], col = "blue")
points(x2[ant$sex == "M"], y2[ant$sex == "M"], col = "red")

pairs(data.frame(x = x2, y = y2, z = z2))
```

## 7. Провести проверку родственных связей. Отфильтровать (при наличии) родственные образцы

```{r}
# TODO check; 100 markers? kinship? etc? maybe duplicates?
dat.f <- snpgdsOpen("DATA_PCA2.gds")
id <- read.gdsn(index.gdsn(dat.f, "sample.id"))
snp <- read.gdsn(index.gdsn(dat.f, "snp.id"))
chr <- read.gdsn(index.gdsn(dat.f, "snp.chromosome"))
pos <- read.gdsn(index.gdsn(dat.f, "snp.position"))
all <- read.gdsn(index.gdsn(dat.f, "snp.allele"))
gen <- read.gdsn(index.gdsn(dat.f, "genotype"))
ant <- read.gdsn(index.gdsn(dat.f, "sample.annot"))

snp.subset <- snpgdsLDpruning(dat.f)
snpset <- as.character(sample(unlist(snp.subset), 300)) # ? TODO try 300
mibd <- snpgdsIBDMLE(dat.f, snp.id = snpset, kinship = T)
snpgdsClose(dat.f)
```
Results:
```{r}
d1 <- snpgdsIBDSelection(mibd, kinship.cutoff = 1 / 8)
dd1 <- d1[order(d1$kinship, decreasing = TRUE),]
dim(dd1) # 220, 6 for N=200; 11, 6 for N=300
```
Filter by IBD:
```{r}
samplesToRemove <- c(d1$ID1, d1$ID2) %>%
  # lapply(function(x) strsplit(x, split = "-")[[1]][2]) %>%
  # lapply(as.integer) %>%
  unique %>%
  unlist %>%
  sort
length(samplesToRemove) # 22 for N=300 // FALSE!!

samplesFilter <- !(id %in% samplesToRemove)

id2 <- id[samplesFilter]
gen2 <- gen[samplesFilter,]
ant2 <- ant[samplesFilter,]
```

Checkpoint:
```{r}
# save.image(file="7_ibd.RData")
load("7_ibd.RData")
```
```{r}
dat.f <- createfn.gds("DATA_FINAL.gds")
add.gdsn(dat.f, "sample.id", val = id2, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.id", val = snp.id, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.chromosome", val = chr, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.position", val = pos, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "snp.allele", val = allele, compress = "ZIP", closezip = TRUE)
add.gdsn(dat.f, "genotype", val = gen2, compress = "ZIP", closezip = TRUE, storage = "bit2", replace = TRUE)
add.gdsn(dat.f, "sample.annot", val = ant2, compress = "ZIP", closezip = TRUE)
put.attr.gdsn(index.gdsn(dat.f, "genotype"), "sample.order")
closefn.gds(dat.f)
```

## 8. Составить отчет о проверке качества данных с указанием составов удаленных образцов или генетических маркеров на каждом из этапов, а также характеристик полученного набора данных для статистического анализа

```{r}
filters.df <- data.frame(
  filter = c("Call rate", "MAF", "% genotyped SNPs", "HWE case", "HWE control", "PCA", "IBD"),
  before = c(ncol(genotype_with_na), # CR
             ncol(genotype_with_na) - sum(crFilter), # MAF
             nrow(genotype_with_na), # % genotyped SNPs
             ncol(genotype_with_na) - sum(crFilter | mafFilter), # HWE case
             ncol(genotype_with_na) - sum(crFilter | mafFilter | hweCaseFilter), # HWE control
             nrow(genotype_with_na) - sum(individFilter), # PCA
             nrow(genotype_with_na) -
               sum(individFilter) -
               sum(!pcaFilter) # IBD
  ),
  after = c(
    ncol(genotype_with_na) - sum(crFilter), # CR
    ncol(genotype_with_na) - sum(crFilter | mafFilter), # MAF
    nrow(genotype_with_na) - sum(individFilter), # % genotypedSNPs
    ncol(genotype_with_na) - sum(crFilter | mafFilter | hweCaseFilter), # HWE case
    ncol(genotype_with_na) - sum(crFilter |
                                   mafFilter |
                                   hweCaseFilter |
                                   hweControlFilter), # HWE control
    nrow(genotype_with_na) -
      sum(individFilter) -
      sum(!pcaFilter), # PCA
    nrow(genotype_with_na) -
      sum(individFilter) -
      sum(!pcaFilter) -
      sum(!samplesFilter) # IBD filter
  )
)
filters.df["delta"] <- filters.df$after - filters.df$before
filters.df["what"] <- c("SNP", "SNP", "samples", "SNP", "SNP", "samples", "samples")

filters.df
```

Итак:

- Изначальные данные: 733 человека (287 женщин + 51 мужчин контрольных, 395 женщин с ВИЧ), 182.558 SNP
- по CallRate (>0.95) и MAF (>0.05) отфильтровали суммарно 35294 SNP (пункт 5)
- по проценту генотипированных SNP (<98%) не отфильтровали никого из индивидов (все качественные)
- по HWE (< 10^{-10} в контрольной группе, < 10^{-5} в исследуемой) прошли все SNP, кроме 1
- по результатам анализа главных компонент отфильтрованы 9 образцов
- по результатам анализа родственных связей (kinship > 1/8) выявили 22 родственников

Итого осталось 727 индивидов, 181.900 SNP (но это с учётом неравновесно сцепленных с фенотипом)

## 9. Провести поиск генетических ассоциаций классическими методами (критерий χ^2/Фишера) при кодоминантной, доминантной и рецессивной альтернативах, а также с использованием аллельного подхода для всех индивидов и только для женской популяции. Построить диаграммы Манхэттена

```{r}
pth.dk <- "DATA_FINAL.gds"
tests <- data.frame(test = NULL, snp = NULL, pv = NULL, stt = NULL)
for (subsetName in c("all", "women")) {
  for (alt in c("cd", "d", "r", "a")) { # co-dominant, dominant, recessive, allelic
    dat.f <- snpgdsOpen(pth.dk, readonly = TRUE)
    pht <- read.gdsn(index.gdsn(dat.f, "sample.annot"))$phenotype
    gen <- read.gdsn(index.gdsn(dat.f, "genotype"))
    snpgdsClose(dat.f)

    if ((subsetName) == women) {
      pht <- pht[pht$sex == "F",]
      pht$sex <- NULL
      gen <- gen[pht$sex == "F",]
    }

    dg <- dim(gen)
    num.p <- dg[1]
    num.g <- dg[2]
    if (alt == "d") {
      gen <- (1 - (gen == 0)) + 2 * (gen == 3) }
    if (alt == "r") {
      gen <- 1 * (gen == 2) + 3 * (gen == 3) }
    if (alt == "a") {
      pht <- c(pht, pht)
      num.p2 <- 2 * num.p
      genotype2 <- matrix(nrow = num.p2, ncol = num.g)
      genotype2[1:num.p,] <- 1 * (gen == 2 | gen == 1) + 3 * (gen == 3)
      genotype2[(num.p + 1):num.p2,] <- 1 * (gen == 2) + 3 * (gen == 3)
      gen <- genotype2
      rm(genotype2) }
    pv <- array(dim = num.g)
    stt <- array(dim = num.g)
    for (j in 1:num.g) {
      gg <- gen[, j]
      gg[gg == 3] <- NA
      tt <- table(gg, pht)
      if (min(dim(tt)) >= 2) {
        if (alt == "cd") {
          tst <- chisq.test(tt)
          pv[j] <- tst$p.value
          ff1 <- !is.na(gg)
          rho <- cor(gg[ff1], pht[ff1]) # Pearson's Rho
          stt[j] <- (1 + rho) / (1 - rho) # Chi squared test statistics
        } else {
          tst <- fisher.test(tt)
          pv[j] <- tst$p.value
          stt[j] <- tst$estimate
        }
      } else {
        pv[j] <- NA
        stt[j] <- NA
      }
    }
    test.tj <- data.frame(test = alt, snp = snp, pv = pv, stt = stt)
    tests <- rbind(tests, test.tj)
  }
  write.table(tests, file = paste0("tests", subsetName, ".csv"), sep = ",", row.names = FALSE)
}
```

Checkpoint:
```{r}
# save.image(file="8_gwas_tests.RData")
load("8_gwas_tests.RData") # in fact this is 9a
```
Manhattan plots:
```{r}
library(qqman)

dat.f <- snpgdsOpen(pth.dk)
snp <- read.gdsn(index.gdsn(dat.f, "snp.id"))
chr <- read.gdsn(index.gdsn(dat.f, "snp.chromosome"))
pos <- read.gdsn(index.gdsn(dat.f, "snp.position"))
snpgdsClose(dat.f)
dat.df <- data.frame(snp = snp, chr = as.numeric(chr), pos = as.numeric(pos))

for (subsetName in c("all", "women")) {
  manh.dat <- read.csv(paste0("tests", subsetName, ".csv"), header = T)
  for (alt in c("cd", "d", "r", "a")) {
    manh.dat.subset <- manh.dat[manh.dat$test == alt,] %>% as.data.frame
    dplyr::left_join(dat.df, manh.dat.subset, by = "snp") %>%
      manhattan(chr = "chr", bp = "pos", snp = "snp", p = "pv", ylim = c(0, 9))
    title(paste("Manhattan:", alt, subsetName))
  }
}
```
## 10. Провести поиск генетических ассоциаций с использованием обобщенной линейной модели логистической регрессии при кодоминантной, доминантной и рецессивной альтернативах, а также с использованием аллельного подхода. Использовать в качестве ковариаты главную компоненту (при необходимости две главных компоненты). Построить диаграммы Манхэттена


```{r}
tests_logreg <- data.frame(test = NULL, snp = NULL, pv = NULL, stt = NULL)
for (subsetName in c("all", "women")) {
  for (alt in c("cd", "d", "r", "a")) { # co-dominant, dominant, recessive, allelic
    dat.f <- snpgdsOpen(pth.dk, readonly = TRUE)
    pht <- read.gdsn(index.gdsn(dat.f, "sample.annot"))$phenotype
    gen <- read.gdsn(index.gdsn(dat.f, "genotype"))
    snpgdsClose(dat.f)

    if ((subsetName) == women) {
      pht <- pht[pht$sex == "F",]
      pht$sex <- NULL
      gen <- gen[pht$sex == "F",]
    }

    dg <- dim(gen)
    num.p <- dg[1]
    num.g <- dg[2]
    if (alt == "d") {
      gen <- (1 - (gen == 0)) + 2 * (gen == 3) }
    if (alt == "r") {
      gen <- 1 * (gen == 2) + 3 * (gen == 3) }
    if (alt == "a") {
      pht <- c(pht, pht)
      num.p2 <- 2 * num.p
      genotype2 <- matrix(nrow = num.p2, ncol = num.g)
      genotype2[1:num.p,] <- 1 * (gen == 2 | gen == 1) + 3 * (gen == 3)
      genotype2[(num.p + 1):num.p2,] <- 1 * (gen == 2) + 3 * (gen == 3)
      gen <- genotype2
      rm(genotype2)
    }

    # PCA
    # dat.f.pca <- createfn.gds(paste0("PCA_", subsetName, alt, ".gds"))
    # add.gdsn(dat.f, "sample.id", val = sample.id, compress = "ZIP", closezip = TRUE)
    # add.gdsn(dat.f, "snp.id", val = snp.id, compress = "ZIP", closezip = TRUE)
    # add.gdsn(dat.f, "snp.chromosome", val = chr, compress = "ZIP", closezip = TRUE)
    # add.gdsn(dat.f, "snp.position", val = pos, compress = "ZIP", closezip = TRUE)
    # add.gdsn(dat.f, "snp.allele", val = allele, compress = "ZIP", closezip = TRUE)
    # add.gdsn(dat.f, "genotype", val = genotype, compress = "ZIP", closezip = TRUE, storage = "bit2", replace = TRUE)
    # add.gdsn(dat.f, "sample.annot", val = sample.annot, compress = "ZIP", closezip = TRUE)
    # put.attr.gdsn(index.gdsn(dat.f, "genotype"), "sample.order")
    # closefn.gds(dat.f)
    #
    # set.seed(0) # for reproducibility
    # ftr <- snpgdsLDpruning(dat.f) # filtering by LD (linkage disequilibrium)
    # ftr.id <- unlist(ftr)
    # q <- snpgdsPCA(dat.f, snp.id = ftr.id)
    # snpgdsClose(dat.f)
    # vec <- q$eigenvect
    # x3 <- vec[, 1]
    # y3 <- vec[, 2]

    pv <- array(dim = num.g)
    stt <- array(dim = num.g)
    for (j in 1:num.g) {
      gg <- gen[, j]
      gg[gg == 3] <- NA
      tt <- table(gg, pht)
      if (min(dim(tt)) >= 2) {

        # logreg.df <- data.frame(pht = pht, gen = gen, pc1)
        model_logreg <- glm(pht ~ as.factor(gg), subset = !is.na(gg), family = "poisson")
        model0 <- glm(pht ~ 1, subset = !is.na(gg), family = "poisson")
        logreg.anova <- anova(model_logreg, model0, test = "LRT")

        pv[j] <- logreg.anova$"Pr(>Chi)"[2]
        stt[j] <- model_logreg$coefficients[2]
      } else {
        pv[j] <- NA
        stt[j] <- NA
      }
    }
    test.tj <- data.frame(test = alt, snp = snp, pv = pv, stt = stt)
    tests_logreg <- rbind(tests_logreg, test.tj)
  }
  write.table(tests_logreg, file = paste0("tests_logreg_", subsetName, ".csv"), sep = ",", row.names = FALSE)
}
```

Manhattan:

```{r}
for (subsetName in c("all", "women")) {
  manh.dat <- read.csv(paste0("tests_logreg_", subsetName, ".csv"), header = T)
  for (alt in c("cd", "d", "r", "a")) {
    manh.dat.subset <- manh.dat[manh.dat$test == alt,] %>% as.data.frame
    dplyr::left_join(dat.df, manh.dat.subset, by = "snp") %>%
      manhattan(chr = "chr", bp = "pos", snp = "snp", p = "pv", ylim = c(0, 9))
    title(paste("Manhattan:", alt, subsetName))
  }
}
```

## 11. Сравнить результаты, полученные классическими методами и с помощью обобщенной линейной модели в присутствии ковариаты

```{r}
```

## 12. Интерпретировать полученные результаты. На уровне значимости 0.1 найти значимые ассоциации по методу Хольма для каждого множественного теста. Помимо значимых проанализировать топ-10 наиболее значимых ассоциаций.

Интерпретация: интересных SNP найдено не было

```{r}

for (subsetName in c("all", "women")) {
  holm.dat <- read.csv(paste0("tests", subsetName, ".csv"), header = T)
  for (alt in c("cd", "d", "r", "a")) {

    holm.dat.subset <- holm.dat[holm.dat$test == alt,]
    holm.dat.subset$padj <- p.adjust(holm.dat.subset$pv, method = "holm")
    holm.dat.subset %>%
      dplyr::arrange(padj) %>%
      head(10) %>%
      print
  }
}

```
